### ISSUE TITLE
Certificate usage cause connection leak and self DOS on Azure

### ISSUE THREAD

**Issue by ayende** [dotnet/corefx#30693]
TLDR, this piece of code:

https://github.com/dotnet/corefx/blob/123b4ed1014df2beb9d36598eff956d8e1f4bb7c/src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertificateAssetDownloader.cs#L85-L93

Can cause high connection usage in some scenarios.

In particular, `DownloadAsset` is called whenever we need to validate the chain on a certificate.

Here is one example stack trace:

```
	System.Security.Cryptography.X509Certificates.dll!Internal.Cryptography.Pal.CertificateAssetDownloader.DownloadAsset(string uri, ref System.TimeSpan remainingDownloadTime)
 	System.Security.Cryptography.X509Certificates.dll!Internal.Cryptography.Pal.CertificateAssetDownloader.DownloadCertificate(string uri, ref System.TimeSpan remainingDownloadTime)
 	System.Security.Cryptography.X509Certificates.dll!Internal.Cryptography.Pal.OpenSslX509ChainProcessor.DownloadCertificate(byte[] authorityInformationAccess, ref System.TimeSpan remainingDownloadTime)
 	System.Security.Cryptography.X509Certificates.dll!Internal.Cryptography.Pal.OpenSslX509ChainProcessor.FindIssuer(System.Security.Cryptography.X509Certificates.X509Certificate2 cert, System.Security.Cryptography.X509Certificates.X509Certificate2Collection[] stores, System.Collections.Generic.HashSet<System.Security.Cryptography.X509Certificates.X509Certificate2> downloadedCerts, ref System.TimeSpan remainingDownloadTime)
 	System.Security.Cryptography.X509Certificates.dll!Internal.Cryptography.Pal.OpenSslX509ChainProcessor.FindCandidates(System.Security.Cryptography.X509Certificates.X509Certificate2 leaf, System.Security.Cryptography.X509Certificates.X509Certificate2Collection extraStore, System.Collections.Generic.HashSet<System.Security.Cryptography.X509Certificates.X509Certificate2> downloaded, System.Collections.Generic.HashSet<System.Security.Cryptography.X509Certificates.X509Certificate2> systemTrusted, ref System.TimeSpan remainingDownloadTime)
 	System.Security.Cryptography.X509Certificates.dll!Internal.Cryptography.Pal.ChainPal.BuildChain(bool useMachineContext, Internal.Cryptography.ICertificatePal cert, System.Security.Cryptography.X509Certificates.X509Certificate2Collection extraStore, System.Security.Cryptography.OidCollection applicationPolicy, System.Security.Cryptography.OidCollection certificatePolicy, System.Security.Cryptography.X509Certificates.X509RevocationMode revocationMode, System.Security.Cryptography.X509Certificates.X509RevocationFlag revocationFlag, System.DateTime verificationTime, System.TimeSpan timeout)
 	System.Security.Cryptography.X509Certificates.dll!System.Security.Cryptography.X509Certificates.X509Chain.Build(System.Security.Cryptography.X509Certificates.X509Certificate2 certificate, bool throwOnException)
 	System.Security.Cryptography.X509Certificates.dll!System.Security.Cryptography.X509Certificates.X509Chain.Build(System.Security.Cryptography.X509Certificates.X509Certificate2 certificate)
 	System.Net.Security.dll!System.Net.Http.TLSCertificateExtensions.BuildNewChain(System.Security.Cryptography.X509Certificates.X509Certificate2 certificate, bool includeClientApplicationPolicy)
 	System.Net.Security.dll!Interop.OpenSsl.AllocateSslContext(System.Security.Authentication.SslProtocols protocols, Microsoft.Win32.SafeHandles.SafeX509Handle certHandle, System.Security.Cryptography.SafeEvpPKeyHandle certKeyHandle, System.Net.Security.EncryptionPolicy policy, bool isServer, bool remoteCertRequired)
 	System.Net.Security.dll!System.Net.Security.SafeDeleteSslContext.SafeDeleteSslContext(System.Net.Security.SafeFreeSslCredentials credential, bool isServer, bool remoteCertRequired)
 	System.Net.Security.dll!System.Net.Security.SslStreamPal.HandshakeInternal(System.Net.Security.SafeFreeCredentials credential, ref System.Net.Security.SafeDeleteContext context, System.Net.Security.SecurityBuffer inputBuffer, System.Net.Security.SecurityBuffer outputBuffer, bool isServer, bool remoteCertRequired)
 	System.Net.Security.dll!System.Net.Security.SecureChannel.GenerateToken(byte[] input, int offset, int count, ref byte[] output)
 	System.Net.Security.dll!System.Net.Security.SecureChannel.NextMessage(byte[] incoming, int offset, int count)
 	System.Net.Security.dll!System.Net.Security.SslState.StartSendBlob(byte[] incoming, int count, System.Net.AsyncProtocolRequest asyncRequest)
 	System.Net.Security.dll!System.Net.Security.SslState.ProcessReceivedBlob(byte[] buffer, int count, System.Net.AsyncProtocolRequest asyncRequest)
 	System.Net.Security.dll!System.Net.Security.SslState.StartReadFrame(byte[] buffer, int readBytes, System.Net.AsyncProtocolRequest asyncRequest)
 	System.Net.Security.dll!System.Net.Security.SslState.StartReceiveBlob(byte[] buffer, System.Net.AsyncProtocolRequest asyncRequest)
 	System.Net.Security.dll!System.Net.Security.SslState.ForceAuthentication(bool receiveFirst, byte[] buffer, System.Net.AsyncProtocolRequest asyncRequest)
 	System.Net.Security.dll!System.Net.Security.SslState.ProcessAuthentication(System.Net.LazyAsyncResult lazyResult)
 	System.Net.Security.dll!System.Net.Security.SslStream.AuthenticateAsServer(System.Security.Cryptography.X509Certificates.X509Certificate serverCertificate, bool clientCertificateRequired, System.Security.Authentication.SslProtocols enabledSslProtocols, bool checkCertificateRevocation)
```

As you can see, this method is called _whenever_ SslStream is used (Kestrel, HTTPS, etc) this will be called.

Now, if you are using Let's Encrypt certificate on Ubuntu, for example, the full certificate chain is not available, so it will download it on every connection.
That issue is already discussed here: https://github.com/dotnet/corefx/issues/26061 and isn't the subject to this issue.

The key here is this code, though:
https://github.com/dotnet/corefx/blob/123b4ed1014df2beb9d36598eff956d8e1f4bb7c/src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertificateAssetDownloader.cs#L85-L93

Skipping the reflection, this is basically:

```
var httpClient = new HttpClient();
return httpClient.GetAsync(uri).Result;
```

And that is a problem, because HttpClient is not meant for use and throw in this manner.  Instead, it is meant to be a long lived object.
This means that it is doing connection pooling.

Now, eventually this will be GCed and finalized. However, in the meantime, if we have a lot of such requests, each of them will spawn a separate connection to the source server.

Now, when you are running on Azure, if your VM doesn't have a public IP, you run through SNAT: https://blogs.msdn.microsoft.com/mast/2015/07/13/azure-snat/

That means that you have about 160 outgoing connections available for the same destination server. 
If you have multiple containers all talking to the same destination server, the number is much lower.
At this point, you basically violate TCP and fun stuff happens. (The destination server get SYN packets for already established connection, in which case you spend 30+ seconds on blocking network wait until the `HttpListener` gives up). 

A possible solution is to cache the `httpClient` instance in `DownloadAsset`, but that has its own risks.  Especially: https://github.com/dotnet/corefx/issues/11224
Not sure whatever that can be ignored at this point.

Workaround is to make sure that the certificate chain can be resolved completely locally, but that isn't always possible and easy to miss.

**Comment by bartonjs** [comment]
@stephentoub

**Comment by stephentoub** [comment]
@ayende, this isn't a new issue you're seeing, right?  This code is new but it replaced direct usage of libcurl, and that direct usage was similarly creating a new connection.

> Now, eventually this will be GCed and finalized. 

The HttpClient is disposed of below the code you highlighted:
https://github.com/dotnet/corefx/blob/123b4ed1014df2beb9d36598eff956d8e1f4bb7c/src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertificateAssetDownloader.cs#L97-L103
What are you seeing being left for finalization?

**Comment by ayende** [comment]
We just run into this in production, where it burned through all the outgoing connection ports on Azure and caused us 30 sec+ hangs that killed the service.

I missed the part about the disposal of the HttpClient, yes.
But that is still a bad practice, because it prevents connection pooling.

**Comment by stephentoub** [comment]
> But that is still a bad practice, because it prevents connection pooling.

I'm not seeing with that.

This code also isn't what you're using in production though, right? i.e. you're using <= 2.1.1, which isn't using HttpClient here at all? Just verifying.

**Comment by ayende** [comment]
Yes, the production code is pre 2.1.1.

**Comment by stephentoub** [comment]
@bartonjs, what do you want to do here?  This issue has been there since 1.0 as far as I can tell.

**Comment by bartonjs** [comment]
Yeah, I'm not sure that there's anything that we could or would do differently.

If the TLS client trusted the root CA and the TLS server sent the intermediates, there'd be no bonus connections.  If the TLS client trusted the root CA, but it had to fetch intermediates, then it would cache the intermediates and not request them again.

If you have a custom callback to work around not trusting the root, a sensible approach could be to do something like

```C#
// do things that can return false before here.  I'm assuming this is a "before next successful call" optimization.

if (!s_cached)
{
    if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
    {
        using (X509Store intermediates = new X509Store(StoreName.CA, StoreLocation.CurrentUser))
        {
            intermediates.Open(OpenFlags.ReadWrite);

            for (int i = 1; i < chain.ChainElements.Length; i++)
            {
                intermediates.Add(chain.ChainElements[i].Certificate);
            }
        }
    }

    s_cached = true;
}
```

That'll add the certs to the intermediates store (even the root cert).  That store is used for lookup, but not trust, and will eliminate the network (modulo CRL checking) for future chain builds.

**Comment by stephentoub** [comment]
I'm still not understanding the "connection leak" part of the issue title.  What's the leak?

> A possible solution is to cache the httpClient instance in DownloadAsset, but that has its own risks. Especially: #11224. Not sure whatever that can be ignored at this point.

#11224 shouldn't be an issue with SocketsHttpHandler, and if it is, the PooledConnectionLifetime can be changed to meet your needs.

But I leave it up to Jeremy as to whether a pool of connections for this specific purpose is a good idea or not.

**Comment by ayende** [comment]
The problem we run into is that we make a lot of requests to the cert provider from Azure.
We run into issue were we started to get hangs, that was because there were enough connections through Azure SNAT that we run out and started reusing the same ports. 
And the other size thought that we had established connection and ignored the SYN packets.

**Comment by karelz** [comment]
@stephentoub @bartonjs can you please help me understand what needs to happen here? Do we need a change in CoreFX?
If yes, is it important scenario (i.e. 3.0)?

**Comment by stephentoub** [comment]
> can you please help me understand what needs to happen here?

Jeremy wrote:
"I'm not sure that there's anything that we could or would do differently."
I don't think anything's changed.  @bartonjs?

**Comment by bartonjs** [comment]
I don't know of new information that would lead to a new conclusion.

There are a couple of different things that could be going on:

* The attempted resources are getting a 404 or other failure.
  * Track an "I've done this recently" state.  (`private static readonly ConcurrentDictionary<string,DateTime> s_notBefore`?)
* The attempted resources are being downloaded, and load as a cert, but are ultimately untrusted, and this keeps happening for the same resources
  * Some sort of weak handle for opportunistic keepalive?
* Lots of different resources are being asked for, they never get asked for again
  * (This is what the code assumes)
  * Now it's a tradeoff between NAT states and longer/queueing requests.  I don't know that we have any justification for changing the current balance of figures.

Since the reported callstack was about setting up the server cert, the answer that I think the Windows Certs team would give (and I'll give) is: just make the server trust its own cert and this problem goes away.  So, as reported, this is a configuration problem.

We could make enhancements for the other problems, but I don't anticipate needing to do them for 3.0.

**Comment by ayende** [comment]
At a minimum, using a single instance of `HttpClient` would help a lot.
Caching that you already have this cert is also possible, but then you get into how much to keep, etc.

**Comment by stephentoub** [comment]
> At a minimum, using a single instance of HttpClient would help a lot.
> Caching that you already have this cert is also possible, but then you get into how much to keep, etc.

Funnily, enough, caching is the same concern for keeping around that HttpClient instance :)

My concern with caching the HttpClient instance would be keeping open a bunch of connections for a long time.  We could mitigate that by setting a short connection lifetime, e.g. a few seconds.  @ayende, how short could we get away with while still mitigating your concerns?  @bartonjs, what would you think of that approach?

**Comment by bartonjs** [comment]
It depends on what short means, I guess.

The chain build has a timeout for how long the operation (in total) is allowed to take.  Most callers don't set it, giving it freedom to do what it wants.  In the degenerate case where there's only one asset to download, the caller said 30 seconds, and the local timeout here is 5, what happens on a 28 second cold-start retrieval?

* 5 seconds, timeout, method returns null, chain is inconclusive, 25 network-seconds were left in the budget.
* 5 seconds, timeout, timeout means retry once, timeout, return null, inconclusive.
* 5 seconds, timeout, repeat until the time limit expires, works after ~28 seconds.
* 5 seconds, timeout, repeat up to 11/12 times (one minute per resource maximum): well, then why wasn't the per-resource timeout just min(one minute, time remaining)?

If short is in the 30-60 second range (i.e. "a short eternity" in computer time) then it's probably fine.  If it's in the sub-10 range then I worry that we're adding intermittency, and sub-10-with-retry sounds like "be the annoying user who hits F5 unnecessarily", but (as someone who occasionally does it to the astonishment of "it's working") maybe it's a "you started but forgot to reply to me" hedge).

**Comment by stephentoub** [comment]
@bartonjs, sorry if I wasn't clear.  By connection lifetime, I didn't mean it would cut a connection short... I meant that when the operation completes and we go to return it to the connection pool, we won't return it if the specified time has been exceeded (or alternatively, we could say we instead want to always return it to the pool but only let it sit in the pool for some max duration).  Today we're creating a connection, using it, and then always immediately destroying it... by storing the HttpClient, we'd then have the ability to say "rather than destroying it, keep it around for a max of N seconds", and we get to choose what that N is.

**Comment by bartonjs** [comment]
@stephentoub Ah.  In that case, I defer to your expertise :smile:.

If the download code is being touched and adding more reflection, it might also be beneficial to do the spiritual equivalent of `client.Timeout = min(client.Timeout, s_cacheTheNonPublicDefaultOf100Seconds)` before calling Get.  (The current "oh, no, we blew the budget!" was from when it was directly using curl)

**Comment by ayende** [comment]
Even a short timeout like 15 seconds would be very useful from my perspective, because under load, it would give us persistent connections that are actively used and the pool will figure out the actual need of the system
